# Movies Tonight Part 3: SQL DML

In [1]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [9]:
%%sql
INSERT INTO MOVIES (MTitle, Rating)
    SELECT DISTINCT MTitle, Rating
    FROM DATASET;
    
SELECT DISTINCT MTitle, Rating FROM DATASET;

 * sqlite:///Movies.db
23 rows affected.
Done.


[('Associate, The', 'PG-13'),
 ('Ghost & The Darkness, The', 'R'),
 ('Independence Day', 'PG-13'),
 ('D3: The Mighty Ducks', 'PG'),
 ('Dear God', ''),
 ('First Wives Club, The', 'PG-13'),
 ('High School High', 'PG-13'),
 ('Larger Than Life', 'PG'),
 ('Mirror Has Two Faces, The', 'PG-13'),
 ('Ransom', 'R'),
 ('Sleepers', ''),
 ('Space Jam', 'PG'),
 ("Stephen King's Thinner", ''),
 ('Michael Collins', 'R'),
 ('Romeo & Juliet', 'R'),
 ('Big Night', 'R'),
 ('Fly Away Home', 'PG'),
 ('Nutty Professor, The', 'PG-13'),
 ('Phenomenon', 'PG'),
 ('Set It Off', 'R'),
 ('Surviving Picasso', 'R'),
 ('Very Brady Sequel, A', 'PG-13'),
 ('That Thing You Do', 'PG')]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [10]:
%%sql
INSERT INTO THEATERS (TName, Location, Phone)
    SELECT DISTINCT TName, Location, Phone
    FROM DATASET;

 * sqlite:///Movies.db
9 rows affected.


[]

In [11]:
%%sql
INSERT INTO ARTISTS (AName)
    SELECT DISTINCT CName
    FROM DATASET;

 * sqlite:///Movies.db
153 rows affected.


[]

In [13]:
%%sql
INSERT INTO CREDITS (MID, AID, CCode)
SELECT DISTINCT MID, AID, CCode
FROM DATASET
    JOIN MOVIES ON (MOVIES.MTitle = DATASET.MTitle)
    JOIN ARTISTS ON (DATASET.CName = ARTISTS.AName);

 * sqlite:///Movies.db
162 rows affected.


[]

In [14]:
%%sql
INSERT INTO SHOWS (MID, TID, STime)
SELECT DISTINCT MID, TID, ShowTime 
FROM DATASET
    JOIN MOVIES ON (DATASET.MTitle = MOVIES.MTitle)
    JOIN THEATERS ON (DATASET.TName = THEATERS.TName);

 * sqlite:///Movies.db
131 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [ ]:
YOUR CODE HERE

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [17]:
%%sql
SELECT DISTINCT COUNT(*)
FROM DATASET;

 * sqlite:///Movies.db
Done.


[(913,)]

In [18]:
%%sql
SELECT COUNT(*)
FROM DATASET;

 * sqlite:///Movies.db
Done.


[(913,)]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

In [19]:
%%sql
DROP TABLE DATASET

 * sqlite:///Movies.db
Done.


[]

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".